File: c++/TripService/TripService.cpp
#ifndef TRIPSERVICE_CPP
#define TRIPSERVICE_CPP

#include "TripService.h"
#include "TripDAO.h"
#include "UserSession.h"
#include <iostream>
#include <list>

using namespace std;

UserSession* UserSession::oneUserSession = 0;

list<Trip> TripService::GetTripsByUser(User* user)
{
    list<Trip> triplist;
    User* loggedUser = UserSession::GetInstance()->GetLoggedUser();
    bool isFriend = false;
    if (loggedUser)
    {
        list<User>::iterator i;
        for (i = user->GetFriends().begin(); i != user->GetFriends().end(); ++i)
        {
            if (*i == *loggedUser)
            {
                isFriend = true;
                break;
            }
        }
        if (isFriend)
        {
            triplist = TripDAO::FindTripsByUser(user);
        }
        return triplist;
    }
    else
    {
        throw "UserNotLoggedInException";
    }
}

void TripService::BuildCheck()
{
    cout << "Built and runnable!" << endl;
}

#endif

File: rust/src/trip_service.rs
use crate::users;

#[derive(Debug, Clone)]
pub struct Trip {}

struct TripDAO {}

impl TripDAO {
    pub fn find_trips_by_user(user: &users::User) -> Result<Vec<Trip>, &'static str> {
        Err("TripDAO should not be invoked on an unit test.")
    }
}

pub struct TripService {}

impl TripService {
    pub fn get_trips_by_user(user: &users::User) -> Vec<Trip> {
        let mut trip_list: Vec<Trip> = vec![];

        let logged_user = &users::ONE_USER_SESSION.get_logged_user().unwrap();

        let mut is_friend: bool = false;

        for i in user.get_friends() {
            if i == logged_user {
                is_friend = true;
                break;
            }
        }

        if is_friend {
            trip_list = TripDAO::find_trips_by_user(user).unwrap();
        }
        trip_list
    }

    pub fn build_check(&self) {
        println!("Built and runnable!");
    }
}
